docs/enhancements/infrastructure-monitoring-service-api/alarms.md

```diff
@@ -0,0 +1,797 @@
+(LIKE alarm_event_record INCLUDING ALL);
+```
+
+Please note that script above was used to test and need some updates for production e.g 
+- use `CREATE TABLE IF NOT EXISTS` and remove `DROP TABLE IF EXISTS`. 
+- do not use * when query 
+- do no use enum and instead go with a lookup table 
+  ```sql
+  CREATE TABLE management_interfaces (
+    id SERIAL PRIMARY KEY,
+    name VARCHAR(255) UNIQUE NOT NULL
+  );
+
+    INSERT INTO management_interfaces (name) VALUES ('O1'), ('O2DMS'), ('O2IMS'), ('OpenFH');
+    CREATE TABLE alarms_table (
+    ...
+    management_interface_id INT REFERENCES management_interfaces(id) NOT NULL DEFAULT 3
+    );
+  ```
+
+Apply this during deployment as specified [here](#k8s-resources) via a migration tool as specified [here](#tooling-and-general-dev-guidelines). 
+
+## Init
+Notes: 
+- Interacting with the following: `versions`, `alarm_dictionary` and `alarm_definitions`
+- `alarm_dictionary` has a primary key `resource_type_id`, ensuring that each ResourceType has a unique AlarmDictionary.
+- There's a one-to-many relation between `alarm_dictionary` and `alarm_definitions` via major.minor OCP version. The referential integrity and normalization version is done through `versions` table
+
+```sql
+INSERT INTO versions (version_number)
+VALUES
+    ('4.16'),
+    ('4.17');
+
+-- Most of the data in this table will come client-go calls
+INSERT INTO alarm_dictionary (alarm_dictionary_version, resource_type_id, entity_type, vendor)
+VALUES
+    ('4.16', 'b3e7149e-d471-4d0f-aaa6-d5e9aa9e713a', 'telco-model-OpenShift-4.16.2', 'Red Hat'),
+    ('4.16', '481688c8-2782-4534-a9de-88ca5154411d', 'telco-model-OpenShift-4.16.8', 'Red Hat' ),
+    ('4.17', 'f8b9e100-fd9f-4923-b96f-89418e9c2560', 'telco-model-OpenShift-4.17.2', 'Red Hat');
+
+-- Insert into alarm_definitions
+INSERT INTO alarm_definitions (alarm_name, alarm_last_change, alarm_description, proposed_repair_actions , alarm_additional_fields)
+VALUES
+    ('NodeClockNotSynchronising','4.16', 'Clock not synchronising.','https://github.com/openshift/runbooks/blob/master/alerts/cluster-monitoring-operator/NodeClockNotSynchronising.md','{"customKey": "customValue"}'),
+    ('NodeClockNotSynchronising','4.17','Clock not synchronising.','https://github.com/openshift/runbooks/blob/master/alerts/cluster-monitoring-operator/NodeClockNotSynchronising.md', '{"customKey": "customValue"}');
+-- The data here mapped from PrometheusRule. See ##PrometheusRule to AlarmDefinition mapping section for more.
+
+-- probable_cause will be auto populated. see the section "##Get Probable cause ID, name and description" for more details
+```
+## Detailed Server Instantiation
+
+For a given OCP release, the alarmDefinitions and probableCauses are fixed, so these can be built up front. For CaaS alarms only one resource type, “NodeCluster”, all alarms map to it.
+
+1. Query all managed clusters to get list of unique major.minor versions
+   - Need to monitor for major.minor new versions
```

Comment:
Yeah that's a great point! Touched on this briefly few weeks ago but for now we can start with this approach for now and maybe move this code resource server eventually? Basically someone has to get the managedCluster CR and read the content....and in this case it's us for now.

But yeah I agree....ultimately Resource server should watch/get cluster resources and we (alarms) should simply be notified about the change and update our db as needed. 

And on the other hand resource server exposes endpoints that needs to return resourceType with its alarm dictionary right? Does that mean we need expose an additional internal GET endpoint for it to retrieve that (assuming we are not sharing the table)? probably a question for https://github.com/openshift-kni/oran-o2ims/pull/262 

cc @bartwensley @browsell @alegacy

Reply:
No, I'm assuming we share the same table. What I meant was to have internal endpoints in the Alarm server so that the Resource server can notify operations on ResourceTypes. For instance, if a new ResourceType is added, then they could send us a Post (with all needed information) for us to determine whether we need to add a new dictionary for this new resource type. If so, then we inspect the managed cluster and get the rules to add a new dictionary. Same for Delete and Update. The idea is to modify the DB accordingly. They still need to go to the DB whenever a query on a resource type is performed. 

What I'm looking with this is to have a more dynamic way to update the dictionaries instead on relying only on restarts.

Reply:
I've updated my [PR](https://github.com/openshift-kni/oran-o2ims/pull/262) with info related to our discussion with respect to the interactions between the two components.